notebooks/04_model_select_and_optimize/hypertune_spark_deep_dive.ipynb

```diff
@@ -1513,57 +1406,57 @@
-       "0  hyperopt  0.930514  0.741341            0.331338   0.306159  247.763664\n",
-       "0    sample  0.980015  0.774216            0.250724   0.230373  399.902761"
+       "   Approach      RMSE       MAE  Explained variance  R squared      Elapsed\n",
+       "0     spark  0.917025  0.728529            0.349054   0.326474   595.752861\n",
+       "0  hyperopt  0.931018  0.745471            0.338381   0.305406  1072.873973\n",
+       "0    sample  0.931756  0.746417            0.337121   0.304305  1034.202902"
       ]
      },
-     "execution_count": 37,
+     "execution_count": 35,
      "metadata": {},
      "output_type": "execute_result"
     }
    ],
    "source": [
-    "comparison"
+    "test_evaluations"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "From the results, it can be seen that, *with the same number of iterations*, Spark native construct based approach takes the least amount of time, and not suprisingly, sample based approach takes the most amount of time. Intuitively this is because Spark native constructs leverage the underlying Java codes for running the actual analytics with high performance efficiency. There is a tradeoff in choosing between `hyperopt` TPE based approach and sampling approach in a for-loop. The former uses TPE for searching optimal parameters intelligently but runs the tuning iterations sequentially, while the latter simple samples parameter values from the pre-defined space but parallelizes the runs with Python `map` function. Run-time performance of the two approaches therefore depends on acutal use cases.  \n",
+    "From the results, it can be seen that, *with the same number of iterations*, Spark native construct based approach takes the least amount of time, even if there is no parallel computing. This is simply because Spark native constructs leverage the underlying Java codes for running the actual analytics with high performance efficiency. Interestingly, the run time for `hyperopt` with TPE algorithm and random search methods are almost the same. Possible reasons for this are that, the TPE algorithm searches optimal parameters intelligently but runs the tuning iterations sequentially. Also, the advantage of TPE may become obvious when there is a higher dimensionality of hyperparameters. \n",
     "\n",
-    "The three approaches use the same RMSE loss. In this measure, the native Spark construct performs the best. The `hyperopt` based approach performs the second best. This may be owing to the limited number of iterations used in the TPE searching algorithm. Note the differences in the RMSE metrics may also come from the randomness of the intermediate steps in parameter tuning process. In practice, multiple runs are required for generating statistically robust comparison results. We have tried 5 times for running the same comparison codes above. The results aligned well with each other in terms of objective metric values and elapsed time. "
```

Comment:
This statement is debatable. What happens is that all three methods perform the same since they get roughly the same validation RMSE (0.94). But they get different RMSE on the test data set. So, the methods do what they are supposed to do but they probably find a bit different hyperparameter values, which lead to some difference on the test data. Or it could be the randomness as you say. So, I don't think that in this experiment hyperopt is worse than Spark. If you tried it with a different dataset you could get better RMSE on test data with hyperopt. It is a matter of statistical variation.

What you can say for sure is that to get very similar statistical performance, spark takes much less time than hyperopt.

Reply:
This is a very good point. Indeed, the overall performance may be from dimensionality of hyperparameters, dataset, randomness of experimentation, etc.